CAIPs/caip-294.md

```diff
@@ -0,0 +1,431 @@
+
+## Test Cases
+
+Here is a test case where we demonstrate a scenario with logic from both a Blockchain Library and a Wallet Provider.
+
+Logic from the Blockchain Library:
+
+```typescript
+// 1. Blockchain Library initializes by listening to wallet_announce messages and
+// also by posting a prompt message
+
+interface WalletMapEntry {
+  params: WalletAnnounceRequestParams;
+  eventName: string;
+}
+
+const wallets: Record<string, WalletMapEntry> = {}
+
+window.addEventListener("caip294:wallet_announce", (event) => {
+  // when an announce message was received then the library can index it by uuid
+  wallets[event.detail.params.uuid] = {
+    params: event.detail.params,
+    eventName: event.detail.params.uuid
+  }
+});
+
+window.dispatchEvent(new CustomEvent(
+  "caip294:wallet_prompt",
+  {
+    detail: {
+      id: 1,
+      jsonrpc: "2.0"
+      method: "wallet_prompt",
+      params: {
+        // optionally the Blockchain Library can prompt wallets to announce matching only the chains
+        chains: []  // optional
+        //  if the Blockchain Library supports CAIP-275 then it can include a name
+        authName: "", // optional
+      },
+    }
+  }
+));
+
+// 2. User presses "Connect Wallet" and the library display the discovered wallets
+
+const selected_uuid = "350670db-19fa-4704-a166-e52e178b59d2"
+
+// 3. User selects a Wallet with UUID = "350670db-19fa-4704-a166-e52e178b59d2" and
+// Blockchain Library will send a CAIP-25 request to establish a wallet connection
+
+const sessionRequest = {
+  id: 123,
+  jsonrpc: "2.0",
+  method: "wallet_createSession",
+  params: {
+    optionalScopes: {
+      eip155: {
+        scopes: ["chain:777"],
+        methods: ["chain_signMessage", "chain_sendTransaction"],
+        notifications: ["accountsChanged"],
+      },
+    },
+    sessionProperties: {
+      expiry: "2024-06-06T13:10:48.155Z",
+    },
+  },
+};
+
+let sessionResult = {}
+
+window.addEventListener(`caip294:${selected_uuid}`, (event) => {
+  if (event.detail.id === sessionRequest.id) {
+    // Get JSON-RPC response
+    if (event.detail.error) {
+      console.error(event.detail.error.message);
+    } else {
+      sessionResult = event.detail.result
+    }
+  }
+});
+
+window.dispatchEvent(new CustomEvent(
+  `caip294:${selected_uuid}`,
+  {
+    detail: sessionRequest
+  }
+));
+
+
+
+// 4. After the response was received by the Blockchain Library from the wallet
+// provider then the session is established with a sessionId matchin the UUID
+// thus signing requests can be using a CAIP-27 request to the wallet user
+const signingRequest = {
+  id: 456,
+  jsonrpc: "2.0",
+  method: "wallet_requestMethod",
+  params: {
+    // UUID from WalletData is used as SessionId
+    sessionId: walletData.uuid,
+    scope: "chain:777",
+    request: {
+      method: "chain_signMessage",
```

Comment:
It depends on your definition of remote... because the idea of CAIP-282 is that it standarizes communication for all transports

That might be window.dispatchEvent, window.postMessage, externally_connectable, etc

Therefore there is always a concept of communication between a webpage and an external party that is remote to the webpage itself

I would not consider remote to be exclusively server-side or onchain

Reply:
That makes sense, I wasn't thinking of them as exclusively as server-side/onchain either. Instead, I was thinking about it more within the context of the different software boundaries that exist and whether we should be leaking them/collapsing them and when.

For example, a page typically interacts with a remote client acting on behalf of the user (say for permission to sign or to retrieve their data) and needs some transport interface to perform that interaction. In most cases, we expect that the page will communicate client via this transport, but because we're utilizing RPCs that don't really align the permissioning model with this (since chains are different/privacy models different) we end up in certain scenarios where the page may choose to bypass interacting with the client and retrieve information directly from the chain instead.

In my mind, this creates a more complex interaction model where it's not clear when the page-client interaction model is expected if the page can access the information itself via the chain which will probably produce odd side effects in a cross chain model.

For this reason, I think it be useful for us to not leak chain specific RPCs to the page and instead should be establishing an encapsulation pattern where the page uses the interface which encapsulates blockchain specific RPC logic away from the page rather than directly calling the RPC methods.

This encapsulation will help with creating a clear interface of reusable patterns to define application logic between the client and page across different blockchain ecosystems such that we can operate under the heuristic of "if the page wants the clients data it should go to the client to get it". Now, I can think of cases where this won't universally apply so I'm not sure it's the cleanest heuristic, but it does seem to be a good practice so that we can establish patterns of reuse between the client and page and make it easier for pages to interact.